### ISSUE TITLE
request: add depth-level argument for [savestate]

### ISSUE THREAD

**Issue by porres** [pure-data/pure-data#1297]
I'm not 100% sure this makes sense, so please bear with me and also help me out.

[savestate] will save the info for a parent patch, we know that.

I'm thinking about [savestate], for instance, inside an abstraction (inside an external, actually) that is loaded inside another abstraction, instantiated twice, let's say.

So, when saving the grandparent, we could have a separate state for each of the two abstractions and the [savestate] object inside the external would load it.

How crazy and undoable is this?

The use case: I'm designing a preset system external abstraction based on [savestate] and [text]. This would allow me/us to easily use this external inside an abstraction.

I do have an alternative way to use this as it is, but it'd be just more elegant and easy if we had this option. 

cheers

**Comment by umlaeute** [comment]
how would you imagine this to actually look on disk?
what happens if you have a `[savestate 1]`  besides an abstraction with a `[savestate 2]`?

how would you keep backwards compatibility?

note that i'm not saying that it is a bad idea: there are just a lot of cases that should be thought about beforehand.

(in practice i think this would require significant overhowl and convolution of the currently very simple and clean implementation; but let's draft a proper proposal first, before getting into the gory details)

**Comment by porres** [comment]
> how would you imagine this to actually look on disk?
> what happens if you have a [savestate 1] besides an abstraction with a [savestate 2]?

"1" could mean "parent" and should be the default, when you don't insert any argument, so it  keeps backwards compatibility. 

Then [savestate 2] would mean "grandparent"

**Comment by umlaeute** [comment]
> Then [savestate 2] would mean "grandparent"


yes oft course, but that's not what I meant.

I meant: how would the state be saved in the (parent) patch?

right now the syntax is:

#A save ....;

**Comment by umlaeute** [comment]
the important part being:
if you currently have multiple `[savestate]` objects in your abstraction, they will all get the same data when the patch is loaded.
this is somewhat ok, as they are all working within the same context (of the patch).

you could of course add some prefixes to the actual persistent data so you can `[route]` it approximately.
but if you have multiple instances of the same child abstraction, 
however, i wonder whether this won't lead to plenty of confusion if you innocently use an object that contains a `[savestate 2]` and you suddenly get weirdo data from your ordinary `[savestate]`.

**Comment by porres** [comment]
> i wonder whether this won't lead to plenty of confusion if you innocently use an object that contains a [savestate 2] and you suddenly get weirdo data from your ordinary [savestate].

I don't think I get this, but sounds like a misusage that should be documented.

My idea is simple. Let me rephrase it. Say you have two abstractions in a parent patch with [savestate], then it gets saved with different data for each... cool, normal behaviour. Now, think that the [savestate] is in an abstraction external used inside the loaded abstraction. The [savestate] with an argument of, say, "2", would just save the data in the grandparent patch instead of the parent patch.

This would allow me to design preset and savestate management eternals.

I think I could do it in an external and test it

**Comment by umlaeute** [comment]
> I don't think I get this, but sounds like a misusage that should be documented.

i don't see how your proposal could be used otherwise (so if the only usage is misusage, we better not).


> My idea is simple. Let me rephrase it. Say you have two abstractions in a parent patch with [savestate], then it gets saved with different data for each... cool, normal behaviour. Now, think that the [savestate] is in an abstraction external used inside the loaded abstraction. The [savestate] with an argument of, say, "2", would just save the data in the grandparent patch instead of the parent patch.

the problem is simple: we currently have a single **slot** to store any number of data along with an object.
this data refers to a single object implicitly (by use of a "last created" logic which will send that data to the last-created object, in our case the abstraction holding the `[savestate]`).

this simple model does not scale to identify multiple concurring `[savestate]` objects that all want to/must share the same single *slot* for storing their data.

this is the reason why currently  only a single `[savestate]` makes much sense, even in a single abstraction: if you have multiple `[savestate]` objects, then they will all receive the same data when the parent patch is restored.
Or, as the help-patch puts it:
> Multiple `[savestate]` objects aren't differentiated, so they all receive all lists sent to any one of them and output them.

for a single abstraction, this is probably OK as they have a well-defined scope (the abstraction itself).

now with your suggestion i see two possibilities:
- either keep the single slot approach, in which case all `[savestate]` instances within the abstraction and it's sub-abstractions with the "correct" depth) will receive the very same data on restore.
- or invent a multi-slot approach where each `[savestate]`  instance only receives the data it sent out.

the latter requires a complete refactoring of the current savestate mechanism, making the thing more complicated and (more importantly) would not be backwards compatible with the current approach.

**The former will leak scope both the the (child) abstractions and the (parent) patch**.
Consider an abstraction for an EQ-effect in the mixer, that has a `gain` property and a `low-cut` property. It will probably want to save the former as a `gain $1` message in the `[savestate]`. Also consider an abstraction for a mixer channel, which has a `pan` property and (by chance) also a `gain` property. It will probably want to save the latter as `gain $1`.

Now, if the *mixer-channel* abstraction is extended to include an *EQ-effect* (with the latter being prepared to save the state on the "parent"), suddenly both will want to save the `gain $1` data. On restore they will both receive both `gain`s, one of them obviously being wrong. How to differentiate?
Also, the *EG-effect* abstraction will receive `pan` data which it doesn't know how to deal with. And the *mixer-channel* abstraction will receive `low-cut` data...

Now, the *mixer-channel* abstraction probably knows that it is using the *EQ-effect* abstraction, and chose to pick a different label for its gain...problem solved (although, isn't the idea of this entire proposal to be able to treat the *EQ-effect* abstraction as a blackbox, and transparently handle whatever it wants to save as its state?).
otoh, the *EQ-effect* abstraction has absolutely no way of knowing in which parent abstraction it is going to be embedded. And really, it should not care (that is the idea of *structured programming*: to not have to worry in which context an abstraction/function/class/... is going to be used; as they are *abstract* solutions for a given problem).

And finally: both single-slot and multi-slot savestates do not tackle the problem of unknown depth at all.
Consider the *EQ-effect* abstraction with a  `[savestate 2]`, so you can easily embed it in your *mixer-channel*.
But now you re-design your mixer with a general *effects* abstraction that bundles a number of different effects (among them the *EQ-effect*), so they can be used not only in the *mixer-channel* but also in the *master-section* and some *aux-module*.  But this means, that the *EQ-effect* now needs a `[savestate 3]` instead of the `[savestate 2]`!!!

No problem, if you design the *EQ-effect* yourself (and are sure that you don't need to save it's state in *different* depths). But what if I just re-use an object from some library?

> I think I could do it in an external and test it.

yes please.
